docs/architecture/blazor-for-web-forms-developers/security-authentication-authorization.md

```diff
@@ -1,15 +1,379 @@
+
+### Roles, claims, and policies
+
+Both the universal providers and ASP.NET Core Identity support the concept of roles. You can create roles for users and assign users to roles. Users can belong to any number of roles, and you can verify role membership as part of your authorization implementation.
+
+In addition to roles, ASP.NET Core identity supports the concepts of claims and policies. While a role should specifically correspond to a set of resources a user in that role should be able to access, a claim is simply part of a user's identity. A claim is a name value pair that represents what the subject is, not what the subject can do.
+
+It is possible to directly inspect a user's claims and determine based on these whether a user should be given access to a resource. However, such checks are often repetitive and scattered throughout the system. A better approach is to define a *policy*.
+
+An authorization policy consists of one or more requirements. Policies are registered as part of the authorization service configuration in the `ConfigureServices` method of `Startup.cs`. For example, the following code snippet configures a policy called "CanadiansOnly" which has the requirement that the user have the Country claim with the value of "Canada".
+
+```csharp
+services.AddAuthorization(options =>
+{
+    options.AddPolicy("CanadiansOnly", policy => policy.RequireClaim(ClaimTypes.Country, "Canada"));
+});
+```
+
+You can [learn more about how to create custom policies in the documentation](https://docs.microsoft.com/aspnet/core/security/authorization/policies).
+
+Whether you're using policies or roles, you can specify that a particular page in your Blazor application require that role or policy with the `[Authorize]` attribute, applied with the `@attribute` directive.
+
+Requiring a role:
+
+```csharp
+@attribute [Authorize(Roles ="administrators")]
+```
+
+Requiring a policy be satisfied:
+
+```csharp
+@attribute [Authorize(Policy ="CanadiansOnly")]
+```
+
+If you need access to a user's authentication state, roles, or claims you can access these by injecting the `AuthenticationStateProvider` into your component:
+
+```csharp
+@using Microsoft.AspNetCore.Components.Authorization
+@inject AuthenticationStateProvider AuthenticationStateProvider
+```
+
+With the provider in place, you can gain access to the user with the following code:
+
+```csharp
+AuthenticationState authState = await AuthenticationStateProvider.GetAuthenticationStateAsync();
+ClaimsPrincipal user = authState.User;
+
+if (user.Identity.IsAuthenticated)
+{
+  // work with user.Claims and/or user.Roles
+}
+```
+
+To work with users and claims you may also need to inject a `UserManager<T>` (use `IdentityUser` for default) which you can use to enumerate and modify claims for a user. First inject the type and assign it to a property:
```

Comment:
Injecting `UserManager<T>` into a component will only work in a Blazor Server app. In a Blazor WebAssembly app, you'd need to provide an endpoint that you can use to get more info about the user or to manipulate the user.

Reply:
Somewhat related, if you want to customize the current user in a Blazor app, typically you would do that using an `AccountClaimsPrincipalFactory`: https://docs.microsoft.com/aspnet/core/blazor/security/webassembly/additional-scenarios#customize-the-user. That's probably a more advanced scenario though than we want to cover here.

Reply:
I agree this only covers Blazor Server, not WebAssembly, but that was by design. The book is all about moving Web Forms developers quickly to Blazor, and seems to focus on the much simpler path of doing so via Blazor Server.

Reply:
Sure, I agree Blazor Server is a simpler path for Web Forms devs, but the Intro, Arch, and Hosting models chapters all talk about Blazor WebAssembly, so it feels a bit incomplete to not even mention that this coding pattern is Blazor Server specific. I think it would be sufficient to just say this pattern is Blazor Server specific, and note that the same pattern won't work in Blazor WebAssembly.

Reply:
Yep - in the latest updates I called it out.